crates/biome_js_analyze/src/lint/nursery/no_restricted_elements.rs

```diff
@@ -0,0 +1,147 @@
+use biome_analyze::context::RuleContext;
+use biome_analyze::{Ast, Rule, RuleDiagnostic, RuleSource, declare_lint_rule};
+use biome_console::markup;
+use biome_deserialize::{
+    Deserializable, DeserializableType, DeserializableValue, DeserializationContext,
+};
+use biome_js_syntax::jsx_ext::AnyJsxElement;
+use biome_rowan::AstNode;
+use rustc_hash::FxHashMap;
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+
+#[cfg(feature = "schemars")]
+use schemars::JsonSchema;
+
+declare_lint_rule! {
+    /// Disallow user JSX elements.
+    ///
+    /// This rule disallows the use of certain custom JSX elements as specified in the rule options.
+    ///
+    /// ## Options
+    ///
+    /// ```json,options
+    /// {
+    ///     "options": {
+    ///         "elements": {
+    ///             "button": "button is not allowed, use Button component instead"
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```jsx
+    /// // {elements: {input: "input is not allowed, use TextField component instead"}}
+    /// <input />
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```jsx
+    /// // {elements: {input: "input is not allowed, use TextField component instead"}}
+    /// <TextField />
+    /// ```
+    pub NoRestrictedElements {
+        version: "next",
+        name: "noRestrictedElements",
+        language: "jsx",
+        sources: &[
+            RuleSource::EslintReact("forbid-elements"),
+        ],
+        recommended: true,
+    }
+}
+
+impl Rule for NoRestrictedElements {
+    type Query = Ast<AnyJsxElement>;
+    type State = CustomRestrictedElementOptions;
+    type Signals = Option<Self::State>;
+    type Options = NoRestrictedElementsOptions;
+
+    fn run(ctx: &RuleContext<Self>) -> Option<Self::State> {
+        let options = ctx.options();
+        if options.elements.is_empty() {
+            return None;
+        }
+        let node = ctx.query();
+        let element_name = node.name().ok()?.to_trimmed_string();
```

Comment:
Don't use `to_trimmed_string` because it allocates a string, and we don't need an allocation in this context. Use `syntax().trimmed_text()`